R/utilities-units.R

```diff
@@ -336,3 +336,200 @@ initializeDimensionAndUnitLists <- function() {
   ospDimensions <<- getDimensionsEnum()
   ospUnits <<- getUnitsEnum()
 }
+
+
+#' Convert data frame to common units
+#'
+#' @description
+#'
+#' When multiple (observed and/or simulated) datasets are present in a data
+#' frame, they are likely to have different units. This function helps to
+#' convert them to a common unit specified by the user.
+#'
+#' This is especially helpful while plotting since the quantities from different
+#' datasets to be plotted on the X-and Y-axis need to have same units to be
+#' meaningfully compared.
+#'
+#' @note
+#'
+#' Molecular weight is **required** for the conversion between certain
+#' dimensions (`Amount`, `Mass`, `Concentration (molar)`, and `Concentration
+#' (mass)`). Therefore, if molecular weight is missing for these dimension, the
+#' unit conversion will fail.
+#'
+#' @return A data frame with measurement columns transformed to have common units.
+#'
+#' @param data A data frame (or a tibble).
+#' @param xUnit,yUnit Target units for `xValues` and `yValues`, respectively. If
+#'   not specified (`NULL`), first of the existing units in the respective
+#'   columns (`xUnit` and `yUnit`) will be selected as the common unit. For
+#'   available dimensions and units, see `ospsuite::ospDimensions` and
+#'   `ospsuite::ospUnits`, respectively.
+#'
+#' @seealso toUnit
+#'
+#' @examples
+#'
+#' # small dataframe to illustrate the conversion
+#' (df <- dplyr::tibble(
+#'   dataType = c(rep("simulated", 3), rep("observed", 3)),
+#'   xValues = c(0, 14.482, 28.965, 0, 1, 2),
+#'   xUnit = "min", xDimension = "Time",
+#'   yValues = c(1, 1, 1, 1, 1, 1),
+#'   yUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   yDimension = c("Amount", "Amount", "Amount", "Mass", "Mass", "Mass"),
+#'   yErrorValues = c(2.747, 2.918, 2.746, NA, NA, NA),
+#'   yErrorUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   molWeight = c(10, 10, 20, 20, 20, 10)
+#' ))
+#'
+#' # default conversion
+#' ospsuite:::.unitConverter(df)
+#'
+#' # customizing conversion with specified unit(s)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h)
+#' ospsuite:::.unitConverter(df, yUnit = ospUnits$Mass$kg)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$s, yUnit = ospUnits$Amount$mmol)
+#'
+#' @keywords internal
+.unitConverter <- function(data, xUnit = NULL, yUnit = NULL) {
+  # target units --------------------------
+
+  # No validation of inputs for this non-exported function.
+  # All validation will take place in the `DataCombined` class itself.
+
+  # The observed and simulated data should have the same units for
+  # visual/graphical comparison.
+  #
+  # Therefore, if target units are not specified by the user, we need to choose
+  # one ourselves. For no special reason, the first element from a vector of
+  # unique units will be selected: one for X-axis, and one for Y-axis, i.e.
+  xTargetUnit <- xUnit %||% unique(data$xUnit)[[1]]
```

Comment:
What about you take the one with the most occurrences?

Reply:
You mean the one with the most number of corresponding rows?
No strong opinion about that. Seems as good as the current strategy to select the first one.

Reply:
If I have min h h h h h h, I would expect hours to be used and not min. 
But not a big deal

Reply:
That's a good idea, but it doesn't fall in the scope of this PR, which mostly about making sure that this function can deal with missing values.

Changing the default logic will also involve changing tests, etc. Better to handle it in a separate PR.